src/serde/ser.rs

```diff
@@ -0,0 +1,804 @@
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_unit_struct(self, _name: &'static str) -> Result<Self::Ok, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_newtype_variant<T: ?Sized>(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: Serialize,
+    {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_tuple(self, _len: usize) -> Result<Self::SerializeTuple, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_tuple_struct(
+        self,
+        _name: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleStruct, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_tuple_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeTupleVariant, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_struct(
+        self,
+        _name: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeStruct, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+
+    fn serialize_struct_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        _variant: &'static str,
+        _len: usize,
+    ) -> Result<Self::SerializeStructVariant, Self::Error> {
+        Err(key_must_be_a_string())
+    }
+}
+
+pub enum TableSerializer<'a, E> {
+    Table(&'a mut Serializer<E>),
+    Timestamp(&'a mut Serializer<E>),
+}
+
+pub struct TimestampSerializer<'a, E> {
+    serializer: &'a mut Serializer<E>,
+}
+
+impl<'a, E> ser::Serializer for TimestampSerializer<'a, E>
+where
+    E: Writer,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    fn serialize_str(self, v: &str) -> Result<Self::Ok, Self::Error> {
+        let datetime = DateTime::parse_from_rfc3339(v).map_err(|e| IonError::EncodingError {
+            description: e.to_string(),
+        })?;
+        let timestamp = Timestamp::from(datetime);
+        self.serializer.writer.write_timestamp(&timestamp)
+    }
+
+    fn serialize_unit_variant(
+        self,
+        _name: &'static str,
+        _variant_index: u32,
+        variant: &'static str,
+    ) -> Result<Self::Ok, Self::Error> {
+        let datetime =
+            DateTime::parse_from_rfc3339(variant).map_err(|e| IonError::EncodingError {
+                description: e.to_string(),
+            })?;
+        let timestamp = Timestamp::from(datetime);
+        self.serializer.writer.write_timestamp(&timestamp)
+    }
```

Comment:
In what case would this be used, out of curiosity? Same goes for `serialize_newtype_struct` and `serialize_some`. Since we control the implementation of `Serialize` for `Timestamp` and we ensure we're passing a `&str` to `serialize_field`, these implementations seems unnecessary as `serialize_str` will certainly always be [the invoked method by `&str`'s implementation of `Serialize`](https://docs.rs/serde/latest/src/serde/ser/impls.rs.html#43-51), no?

Reply:
Most likely true, I was just copying what toml-rs did, they had these implemented as well intriguingly enough. Not exactly sure why.

Reply:
Ok roger that. Doesn't affect the implementation of course, but just for someone like me reading the code, I'd be yearning for a comment to understand in what case it was necessary. I think it's best to return an error, since these methods are not expected to be invoked.